_posts/2016-08-10-dynamic-loading.md

```diff
@@ -0,0 +1,91 @@
+virtual addresses into a physical address in real memory, which is shared
+between all processes. Unfortunately, in the world of embedded systems MMUs are
+not available. Processors like the ARM Cortex-M series omit them because they
+aren't necessary for traditional single-process embedded devices.
+
+A common approach to handle this issue is to assign addresses to applications
+at compile-time. For example, Application A can be placed at address 0x21000
+and Application B can be placed at address 0x22000. This runs into problems
+when Application A grows in size in the future. Moreover, with Tock we want to
+be able to handle dynamically adding, updating, and removing applications at
+run time. Assigning each an address in advance simply isn't possible.
+
+In Tock, we use position independent code (PIC) [^1] to enable loading
+multiple applications. In PIC, all branches and jumps are PC-relative
+rather than absolute, allowing code to be placed at any address in
+Flash. All references to the data section are indirected through the
+Global Offset Table (GOT) [^2]. Rather than access data at an absolute
+address, first the address of the data is loaded from a hard-coded
+offset into the GOT, and then the data is accessed at that address. This
+allows the OS to simply relocate all addresses in the GOT at load time
+based on the actual location of SRAM rather than fixing various
+instructions throughout the code. The address of the GOT itself is
+stored in a PIC base register which is set by the OS before switching to
+application code and is different for each application.
+
+While PIC handles the majority of addressing issues, it does not fix
+everything. Data members which are themselves pointers are assigned a
+value by the compiler rather than indirecting through the GOT. For
+example, take the statement:
+
+```c
+  char* str = "Hello";
+```
+
+The address of `"Hello"` is stored in the data section (as the value of
+`str`) and is not relocated like elements of the GOT. In order to solve
+this issue, we collect the addresses of necessary relocations (such as
+the address of `"Hello"`) from the ELF and append them to the binary.
+Tock can then fix each address at load time based on the actual location
+of the application’s text and data segments. Since the application is
+already compiled as PIC, remaining fixes will only exist in the data
+segment, making the process of applying relocations simple.
+
+Combining these two solutions together, we reach the Tock application
+format. Each application binary is compiled as position independent
+code [^3], has a relocation section appended to it [^4], and begins with a header
+structure containing the size and location of the text, data, GOT, and
+relocation segments as well as an entry point for the app [^5]. The app is
+able to be loaded into any flash and SRAM addresses with the recurring
+cost of additional load instructions for the indirection through the GOT
```

Comment:
This phrasing makes the cost sound very expensive. It reality the ARM ISA is heavily optimized for PIC. I think in many cases it costs either nothing or at most one additional LDR (2 cycles)